File: backend/functions/src/features/authentication/domain/extensions/user_entity_reference.ts
import type {
  CollectionReference,
  DocumentData,
  DocumentReference,
  Firestore,
} from "firebase-admin/firestore";

/** Espelho de `UserEntityReference` do app (Firestore + metadados de cache). */
export const UserEntityReference = {
  firebaseUidReference(
    firestore: Firestore,
    uid: string,
  ): DocumentReference<DocumentData> {
    return UserEntityReference.firebaseCollectionReference(firestore).doc(uid);
  },

  firebaseCollectionReference(
    firestore: Firestore,
  ): CollectionReference<DocumentData> {
    return firestore.collection("Users");
  },

  cachedKey(): string {
    return "CACHED_USER_INFO";
  },

  userFields: ["email", "password", "phoneNumber"] as const,
};

File: app/lib/features/users/domain/extensions/user_entity_reference.dart
import 'package:app/features/users/domain/entities/user_entity.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

extension UserEntityReference on UserEntity {
  static DocumentReference<Map<String, dynamic>> firebaseUidReference(
    FirebaseFirestore firestore,
    String uid,
  ) {
    return firebaseCollectionReference(firestore).doc(uid);
  }

  static CollectionReference<Map<String, dynamic>> firebaseCollectionReference(
    FirebaseFirestore firestore,
  ) {
    return firestore.collection('Users');
  }

  static String cachedKey() => 'CACHED_USER_INFO';

  static final List<String> userFields = [
    'email',
    'password',
    'phoneNumber',
  ];
}
